keps/0022-diagnostics-bundle.md

```diff
@@ -0,0 +1,388 @@
+- Information about the KUDO instance's Operator and OperatorVersion
+- Logs for the KUDO controller manager
+- Describe for the KUDO controller manager resources
+- RBAC resources that are applicable to the KUDO controller manager
+- Current settings and version information for KUDO
+- k8s events (can we filter them for resources that the instance owns?)
+
+Operator developer experience, then, focuses on customizing diagnostics
+information to gather information about the running application. The following
+forms are available, subject to change over time:
+
+- **Copy**: Copy a file out of a running pod. This is useful for non-stdout/err
+  logs, configuration files, and other artifacts generated by an application.
+  Higher level resources can also be used, which will copy the file on all pods
+  selected by that resource.
+- **Command**: Run a command on a running pod and copy the stdout/err. Higher level
+  resources can also be used, which will run the command on all pods selected by
+  that resource.
+- **Task**: Run a KUDO task and copy the stdout and other arbitrary files.
+- **HTTP**: Make an HTTP request from the KUDO controller manager to a named
+  service and port and copy the result of the request.
+
+While some of these are redundant (HTTP can be a command or job), the intent
+is to provide a high level experience where possible so that operator developers
+don't necessarily need to maintain a `curl` container as part of their
+application stack.
+
+Operator-defined diagnostics collection is defined in a new `diagnostics.bundle.resources`
+key in `operator.yaml`:
+
+```yaml
+diagnostics:
+  bundle:
+    resources:
+      - description: Zookeeper Configuration File
+        name: "zookeeper-configuration"
+        kind: Copy
+        spec:
+          path: 
+            - /opt/zookeeper/server.properties
+            - /opt/zookeeper/zoo.cfg
+          selector:
+            matchLabels:
+              app: zookeeper
+              heritage: kudo
+          objectRef:
+            kind: StatefulSet # Runs on ALL pods in the statefulset
+      - description: DNS information for running pod
+        name: "dns-information"
+        kind: Command
+        spec:
+          command: # Can be string or array
+            - nslookup
+            - google.com
+          selector:
+            matchLabels:
+              app: zookeeper
+              heritage: kudo
+          objectRef:
+            kind: Pod
+    filters:
+      - name: Authentication information
+        spec:
+          regex: "^host: %w+$"
+```
+
+This key is **OPTIONAL**. Default diagnostics collection will happen regardless
+of the `diagnostics.bundle` key's presence. Note, moving to a graph-based engine
+for KUDO will make selecting of resources much easier, rather than having to
+use magical strings with templates. Future iterations of this will reduce the
+complexity of selecting resources to run commands and files on.
```

Comment:
```suggestion
of the `diagnostics.bundle` key's presence. Future iterations of this might reduce the
complexity of selecting resources to run commands and files on.
```